library/ssl_tls.c

```diff
@@ -2828,15 +2829,28 @@ int mbedtls_ssl_handshake_step( mbedtls_ssl_context *ssl )
         MBEDTLS_SSL_DEBUG_MSG( 2, ( "client state: %s",
                                     mbedtls_ssl_states_str( ssl->state ) ) );
 
-#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
-        if( mbedtls_ssl_conf_is_tls13_only( ssl->conf ) )
-            ret = mbedtls_ssl_tls13_handshake_client_step( ssl );
-#endif /* MBEDTLS_SSL_PROTO_TLS1_3 */
+        switch( ssl->state )
+        {
+            case MBEDTLS_SSL_HELLO_REQUEST:
```

Comment:
Note (again, pre-existing, out-of-scope): last time I looked at it, I couldn't understand what this state is for. Did you y any chance understand why it exists? If not, we should try removing it at some point, it's not good to have in the code base things that we don't understand but dare not touch.

Reply:
I guess it is related to the Hello Request message defined in RFC 5246. I wanted to get rid of it but saw this message in the RFC thus stopped thinking about it. It is maybe the right thing to do at some point though.

Reply:
Yes, IIRC the Hello Request message is not really part of the handshake, it's used by the server to request that the client starts a new handshake (renegotiation), so it hardly makes sense for it to be the starting point of non-renegociated hanshakes, especially client-side, and I'm not sure it makes sense for it to be a handshake state at all. Anyway, orthogonal to this PR again, so marking as resolved.